### PR TITLE
const Path initialisation

### PR DESCRIPTION
**Pull request by conradludgate**
`Path::new` cannot be made `const` currently, due to its usage of `AsRef<OsStr>`.

Since Rust does not yet have any concept of const traits, this PR adds a couple new methods in order to construct const Paths.

The expected usage is

```rust
const PATH: &Path = Path::from_os_str(OsStr::from_str("/my/path"));
```

## Alternatives

In some future, we may get const traits, such that we could do
```rust
impl const From<&str> for &Path {
    // ...
}

const PATH: &Path = Path::from("/my/path");
```

### CODE REVIEW

**Comment by joshtriplett** [comment]
I think it may *sometimes* make sense to have methods like this regardless, both for self-documenting code and to usefully constrain inference.

To evaluate whether these methods carry enough weight to add, it would help to know roughly when enough of const traits will become available to make these methods unnecessary for that purpose.

cc @oli-obk @rust-lang/wg-const-eval; do you have a *rough* idea of the timeline upon which we can support const traits well enough to not need these methods? That would help determine whether we want to add workarounds like this.

**Comment by fee1-dead** [comment]
Perhaps a sane way is to stabilize a subset of `const_trait_impl` that only allows *calling* functions from concrete `const` implementations of a trait. We could save *defining* a `impl const`, calling/defining a function that uses const trait bounds for stabilization later.

**Comment by joshtriplett** [comment]
This looks reasonable to me.

It seems like the common case will be making a const path from a `str`, rather than an `OsStr`. I think it would make sense to add a method for doing that directly. Such a method probably shouldn't use the name `from_str` since that would conflict with `FromStr::from_str`, but I think a method should exist for that to simplify the common case.

cc @rust-lang/libs-api: name suggestions?

Also, could you please file a tracking issue and use it in the stability attributes?

r=me with those changes.

File: library/std/src/path.rs [hunk 1]
```diff
@@ -1973,7 +1973,27 @@ impl Path {
...
+    /// #![feature(const_path)]
+    /// use std::ffi::OsStr;
+    /// use std::path::Path;
+    ///
+    /// const PATH: &Path = Path::from_os_str(OsStr::from_str("/foo/bar"));
+    /// ```
+    #[inline]
+    #[unstable(feature = "const_path", reason = "TBD", issue = "none")]
+    pub const fn from_os_str(s: &OsStr) -> &Path {
+        unsafe { &*(s as *const OsStr as *const Path) }
```
Comment (the8472) [inline library/std/src/path.rs:1996]: `osstr.as_ref() -> Path` already exists. Imo it's better to wait until that can be made const before introducing additional API surface for the sole purpose of supporting constification.

**Comment by pitaj** [comment]
Ping from triage. @conradludgate you still need to reference the tracking issue in your stability attributes.

@joshtriplett It appears there are mixed reactions, are you still interested in these additions?

**Comment by JohnCSimon** [comment]
@conradludgate 

Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you are going to continue please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

**Comment by ferdonline** [comment]
I find this really useful. Without it we sometimes have to resort to `lazy_static`. If the community is interested I can help pushing this forward.

**Pull request closed without merging**